solutions/cf-1557B.mdx

```diff
@@ -0,0 +1,345 @@
+</center>
+
+Thus, we need 4 subarrays at minimum.
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+<Spoiler title="Time Complexity Proof">
+Sorting Requires $\mathcal{O}(N\log N)$, and repeated map insertions also require $\mathcal{O}(N\log N)$. Hence the overall time complexity is $\mathcal{O}(N\log N)$.
+</Spoiler>
+	
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+typedef long long ll;
+#define pb push_back
+#define fastio ios::sync_with_stdio(0); cin.tie(0)
+#define MOD 1000000007
+#define FF first
+#define SS second
+ 
+void solve() {
+	int n, k; cin >> n >> k;
+	int arr[n];
+	int sarr[n];
+	for (int i = 0; i < n; i++) {
+		cin >> arr[i];
+		sarr[i] = arr[i];
+	}
+	map<int, int> next; // Map from one value to the value after it in the sorted array
+	sort(sarr, sarr + n);
+	for (int i = 0; i < n - 1; i++) {
+		next[sarr[i]]=sarr[i+1];
+	}
+	next[sarr[n-1]] = MOD; // Ridiculously Large Number
+	int ans = 1;
+	for (int i = 1; i < n; i++) {
+		// If the previous element in the unsorted array is not the same as the previous element in the sorted array, we need at least 1 more segment
+		if (next[arr[i-1]]!=arr[i]) ans++; 
+	}
+	if (ans <= k) cout << "YES\n";
+	else cout << "NO\n";
+}
+ 
+int main() {
+	fastio;
+	int t; cin >> t; while (t--) solve();
+}
+```
+</CPPSection>
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+
+public class MoamenSubarrays {
+	public static void main (String[] args) throws IOException {
+		BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
+		StringTokenizer st = new StringTokenizer(f.readLine());
+	
+		int testNum = Integer.parseInt(st.nextToken());
+		for (int t = 0; t < testNum; t++) {
+			st = new StringTokenizer(f.readLine());
+			int n = Integer.parseInt(st.nextToken());
+			int k = Integer.parseInt(st.nextToken());
+			
+			int[] arr = new int[n];
+			int[] sarr = new int[n];
+	
+			st = new StringTokenizer(f.readLine());
+			for (int i = 0; i < n; i++) {
+				arr[i] = Integer.parseInt(st.nextToken());
+				sarr[i] = arr[i];
+			}
+			Arrays.sort(sarr);
+					      
+			TreeMap<Integer, Integer> next = new TreeMap<>(); // Map from one value to the value after it in the sorted array
+			for (int i = 0; i < n - 1; i++) {
+				next.put(sarr[i], sarr[i + 1]);
+			}
+			next.put(sarr[n - 1], (int) 1e9);
+								       
+			int ans = 1;
+			for (int i = 1; i < n; i++) {
+				// If the previous element in the unsorted array is not the same as the previous element in the sorted array, we need at least 1 more segment
+				if (next.get(arr[i - 1]) != arr[i]) ans++;
+			}
+	
+			if (ans <= k) System.out.println("YES");
+			else System.out.println("NO");
+		}
+	}
+}
+
+```
+</JavaSection>
+	
+<PySection>
+```python
+for _ in range(int(input())):
+	n, k = map(int, input().split())
+	arr = list(map(int, input().split()))
+
+	sarr = arr.copy()
+	sarr.sort()
+	
+	next = {} # Map (dict) from one value to the value after it in the sorted array
```

Comment:
ok so for these inline comments, it's the convention the leave two spaces after the line of code then put the comment
so
```suggestion
	next = {}  # Map (dict) from one value to the value after it in the sorted array
```
also it's generally unadvisable to name a variable that overrides a builtin function even if you aren't using it

Reply:
Oh yeah forgot about next 😐